### DISCUSSION TITLE
step-ca breaks after some time

### DISCUSSION THREAD

**Discussion by staxfur** [smallstep/certificates#598] [Q&A]
Hello,
I have step-ca running in  a docker container inside of an lxc and faced the problem that one of my reverse proxies (caddy) tried to order certificates for an internal domain which is not set up in my DNS server. The result was that step-ca created 100% cpu usage in my lxc and was not responding anymore (the health check was still passing, but certificate orders and renews were not possible anymore.) As I restarted the lxc, step-ca was functioning normally again. Some days later the same problem (with an different caddy that started ordering for one or two days). This time I restarted the step-ca lxc again but the docker container was refusing to start because it returned that there is a problem with the badger database:
`value log truncate required to run db. this might result in data loss`
The database was over 200mb in size and as I have found nothing via a google search, I have simply deleted the database to let step-ca re-create it. Apparently step-ca started and I could order certificates for my hypervisor, but as I started my caddy reverse proxy, step-ca complained that it had read/write errors in the badger database file. I restarted it and not even starts since then as it immediately shows an error that it can't write into the badger database...

My question is why is step-ca / badger broken just because of many certificate requests? And how do I prevent those errors? Or are there other recommendations about hosting step-ca in a "production environment"? (in quotes because it's just for my private and local infrastructure)

**Comment by tashian** [comment]
Hi @MCWertGaming hard to say if this is a step-ca issue or some interaction between step-ca and caddy. It certainly seems like a step-ca issue in how we are handling requests for domains that don't exist.
Is it clear from the step-ca logs that step-ca is getting into a loop?
Can you share a portion of your step-ca logs where the ACME challenge and/or DNS lookup is failing?
I'm also wondering if running it in a container is making any difference in behavior here.

In terms of "production" environment with lots of requests, generally speaking a single step-ca instance can handle a lot of ACME traffic and it doesn't require a lot of CPU to issue certificates. I'd suggest checking out our [Production Considerations](https://smallstep.com/docs/step-ca/certificate-authority-server-production) documentation.

**Reply by staxfur** [reply]
Hi! Thank you for your reply. I'm really sure that this is a step-ca related issue, because the step-ca process is freaking out. The strange thing is that it seems like nothing is wrong, i have only noticed it because my certificates weren't renewed today. But the lxc had 100% CPU usage and was not responding (lxc guest and step-ca) and like I said, the ca.local/health page was still visible and was showing "Status: OK".

So it would be really strange if the issue was related to caddy. And even if it was caused by caddy I would say that step-ca would at least be secured against such. I mean like just imagine someone (maybe me) would be opening the CA to the public internet and someone would know of this issue and starts to crash my ca. That would be really bad.

Since I can't get informations out of the container (as I have a systemd service that starts the sontainer with docker and automatically re-creates it on startup I can't read the docker or systemd logs), I'll setup the ca again and enable loggin in step-ca. As the error came up after something like a few days, I'll check on wether this happens again and come back to this thread when the ca broke again. I just thought that this issue would be already known. 

Thanks for the Help anyway!

**Reply by staxfur** [reply]
Well, is there a way to let step-ca log into a file or something like that itself? Can't find it in the documentation. Otherwise I could just write the command output into a file...

**Reply by staxfur** [reply]
Also: What is the maximum size of the badger db? I think that that was the problem after all.

**Reply by tashian** [reply]
Thanks for looking deeper into this. I'd love to be able to reproduce the issue, but I need more information about what's happening in the CA. I'm not sure if there's a max size for a Badger DB, I couldn't find any reference to a max size in their documentation. It sounds like that could be an issue though. I'm also wondering if the container volume might be maxing out? I guess we'll learn more from the logs.

step-ca outputs logs to STDERR, so you can run something like `step-ca config/ca.json 2> log` to output the log to a file.

In addition to issued certificate metadata, the Badger DB stores open ACME orders. It will fill up quickly if there are many, many open orders that haven't been completed. My hunch is that that's the case here. For reference, I've been running step-ca to get a handful of certs in my homelab, for several months, and the DB size is 7MB.

**Reply by staxfur** [reply]
The container Volume had 8 gigabytes in storage (system + docker + step-ca are only 1G so it's more than enough.) I have logged a bit now and my caddy is flodding my ca with certificate requests (8 per second) and I can watch how the storage of my lxc increases in size. Caddy is creating that much requests because the ca can't validate it due to a wrong dns name.

I think that it would be a good thing to simply limit the requests for a domain, like only 5 open requests total and dropping old ones after maybe 5 minutes or something like that.

Here is a screenshot:
![image](https://user-images.githubusercontent.com/39553804/120679267-858ae200-c488-11eb-948b-62c1933ec9b1.png)

Here is a whole line:
```
time="2021-06-03T15:52:45Z" level=info duration=5.724886ms duration-ns=5724886 fields.time="2021-06-03T15:52:45Z" method=POST name=ca nonce=MEtjd0FBMHZxTTlkUm1C path=/acme/acme/authz/mH451EBDDxzfdQ protocol=HTTP/2.0 referer= remote-address=XX.XX.XX.XX request-id=c2sflj9a response="{\"identifier\":{\"type\":\"dns\",\"value\":\"domain.local\"},\"status\":\"pending\",\"challenges\":[{\"type\":\"dns-01\",\"status\":\"pending\",\"token\":\"kXrrHrykQf8j5WqID3zj\",\"url\":\"https://ca.local/acme/acme/challenge/mH451EBf/W7pczBN\"},{\"type\":\"http-01\",\"status\":\"pending\",\"token\":\"kXrrHrykQ\",\"url\":\"https://ca.local/acme/acme/challenge/mH451EBDDxzHDMnxGkxQY\",\"error\":{\"type\":\"urn:ietf:params:acme:error:connection\",\"detail\":\"The server could not connect to validation target\"}},{\"type\":\"tls-alpn-01\",\"status\":\"pending\",\"token\":\"kXrrHry8JN4K62j\",\"url\":\"https://ca.local/acme/acme/challenge/mH451EBDDxzfd"}],\"wildcard\":false,\"expires\":\"2021-06-04T15:51:34Z\"}" size=882 status=200 user-agent="Caddy/2.4.1 CertMagic acmez (linux; amd64)" user-id=
```
(I have removed IPs, tokens and domains)

I'll Create an Issue on the caddy github repository as well as it shouldn't try to order 8 times a second as long as it fails...

Tell me if you need further informations and thank you really much for helping me out with this issue!

**Reply by staxfur** [reply]
Small addition: My step-ca machine started with a total ram usage of 1.05 gigabytes, then it increased to 1.10 gigabytes in roughly about 1.5h. Now (another ~1.5h later) it's back at 1.07 gigabytes. Interesting, I'll keep an eye on that the next days!

**Reply by tashian** [reply]
See also caddyserver/caddy#4186 for details of the Caddy side of things (potential issue with Caddy's ACME client).

**Reply by tashian** [reply]
Just opened #601 to look further into this on the step-ca side. Thanks Damon.

**Reply by staxfur** [reply]
Here is a demonstration of the caddy side https://github.com/caddyserver/caddy/issues/4186#issuecomment-854134793. Maybe that helps in any way.
